### ISSUE TITLE
Text value for text nodes?

### ISSUE THREAD

**Issue by nemo** [oscarotero/node-sketch#7]
Hey – been messing around with this module a bit for the past couple of days (awesome job, btw!)

Haven't figured out how to get the actual text value of an element though. I only see things like this:

```
{ _archive: 'YnBsaXN0MDDUAQIDBAUGZGVYJHZlcnNpb25YJG9iamVjdHNZJGFyY2hpdmVyVCR0b3ASAAGGoK8QGwcIDxAcHR4fKywtLi8wNjpCQ0RFRklNV1pdYFUkbnVsbNMJCgsMDQ5YTlNTdHJpbmdWJGNsYXNzXE5TQXR0cmlidXRlc4ACgBqAA1tIZWxsbyAjdGV4dNMREgoTFxtXTlMua2V5c1pOUy5vYmplY3RzoxQVFoAEgAWABqMYGRqAB4AOgBaADV8QKk1TQXR0cmlidXRlZFN0cmluZ0NvbG9yRGljdGlvbmFyeUF0dHJpYnV0ZV8QH01TQXR0cmlidXRlZFN0cmluZ0ZvbnRBdHRyaWJ1dGVfEBBOU1BhcmFncmFwaFN0eWxl0xESCiAlG6QhIiMkgAiACYAKgAukJiYmJoAMgAyADIAMgA1TcmVkVWFscGhhVGJsdWVVZ3JlZW4jP/AAAAAAAADSMTIzNFokY2xhc3NuYW1lWCRjbGFzc2VzXE5TRGljdGlvbmFyeaIzNVhOU09iamVjdNIKNzg5XxAaTlNGb250RGVzY3JpcHRvckF0dHJpYnV0ZXOAFYAP0xESCjs+QaI8PYAQgBGiP0CAEoATgBRfEBNOU0ZvbnRTaXplQXR0cmlidXRlXxATTlNGb250TmFtZUF0dHJpYnV0ZSNASAAAAAAAAF8QEk1vbnRzZXJyYXQtUmVndWxhctIxMkdIXxATTlNNdXRhYmxlRGljdGlvbmFyeaNHMzXSMTJKS18QEE5TRm9udERlc2NyaXB0b3KiTDVfEBBOU0ZvbnREZXNjcmlwdG9y1U5PClBRUlNUVVVaTlNUYWJTdG9wc1tOU0FsaWdubWVudFxOU1RleHRCbG9ja3NbTlNUZXh0TGlzdHOAABAEgBmAF4AX0hIKWFmggBjSMTJbXFdOU0FycmF5ols10jEyXl9fEBBOU1BhcmFncmFwaFN0eWxlol410jEyYWJfEBJOU0F0dHJpYnV0ZWRTdHJpbmeiYzVfEBJOU0F0dHJpYnV0ZWRTdHJpbmdfEA9OU0tleWVkQXJjaGl2ZXLRZmdUcm9vdIABAAgAEQAaACMALQAyADcAVQBbAGIAawByAH8AgQCDAIUAkQCYAKAAqwCvALEAswC1ALkAuwC9AL8AwQDuARABIwEqAS8BMQEzATUBNwE8AT4BQAFCAUQBRgFKAVABVQFbAWQBaQF0AX0BigGNAZYBmwG4AboBvAHDAcYByAHKAc0BzwHRAdMB6QH/AggCHQIiAjgCPAJBAlQCVwJqAnUCgAKMApkCpQKnAqkCqwKtAq8CtAK1ArcCvALEAscCzALfAuIC5wL8Av8DFAMmAykDLgAAAAAAAAIBAAAAAAAAAGgAAAAAAAAAAAAAAAAAAAMw' }
```

Inside an attribute called `attributedString` on the text node.

Any thoughts on how to get the actual text value? That looks like a dump from an Objective C object into JSON and not sure what format it's in..

**Comment by oscarotero** [comment]
Sadly, the sketch file format is not entirely opened, there're things that are not human readable like text layers, which seems described as BASE64 encoded Binary Plists, pasted inside the JSON files.
I wasn't able to read/write to this format, but I'm open to contributions. If you found a way to handle these values, please let me know.

**Comment by nemo** [comment]
@oscarotero found a super hacky way to handle them for now. Ended up realizing what you have: BASE64 encoded Binrary PLists.

the issue with them is how shaky it is to grab the values out of them (array index), and lots of indexes for things like text alignment etc.

Over-time I'm fairly certain I'm going to have it all mapped out, so I'll be sure to slowly send PRs the moment I have parts of it mapped out and feel comfortable having done so.

**Comment by nemo** [comment]
For context – I'm using this in production, so we'll certainly need it to work properly :)

**Comment by oscarotero** [comment]
Good news! 🎆
I'm looking forward to seeing it.

**Comment by salmon-charles** [comment]
Hello, 

I am not sure how you would like to integrate it into your lib so please find the following code as a contribution. 
It works for me.

```javascript

const ns = require('node-sketch');
const bplist = require('bplist-parser');

ns.read('file.sketch').then(sketch => {
    sketch.textStyles.forEach((node) => {
            var textStyle = node.get('style').get('textStyle');
            var fontAttribute = getArchive(textStyle.encodedAttributes.MSAttributedStringFontAttribute);
    })
})

function getArchive(node) {

    var stringArchive = node._archive;
    const buf = Buffer.from(stringArchive, 'base64');
    var archive = bplist.parseBuffer(buf);

    let result = {};

    let objects = archive[0].$objects;

    let root = archive[0].$top.root.UID;

    let getReferenceById = (id) => {
        let r = {};
        let o = objects[id];
        if (typeof o === "string" || typeof o === "number" || typeof o === "boolean") {
            return o;
        }


        if (typeof o === "object") {
            for (var i in o) {
                if (o[i].UID) {
                    r[i] = getReferenceById(o[i].UID);
                } else if (Array.isArray(o[i]) && i !== "NS.keys" && i !== "NS.objects") {
                    r[i] = [];
                    o[i].forEach((ao) => {
                        if (ao.UID) {
                            r[i].push(getReferenceById(ao.UID));
                        } else {
                            r[i].push(ao);
                        }
                    });
                } else if(i !== "NS.keys" && i !== "NS.objects") {
                    r[i] = o[i];
                }

            }
        }

        if (o['NS.keys']) {
            o['NS.keys'].forEach((keyObj, index) => {
                let key = getReferenceById(keyObj.UID);
                let obj = getReferenceById(o['NS.objects'][index].UID);
                r[key] = obj;
            });
        }
        return r;
    };


    let topObj = objects[root];
    for (var key in topObj) {
        if (topObj[key].UID) {
            result[key] = getReferenceById(topObj[key].UID);
        }
    }
    return result;
}

```

**Comment by oscarotero** [comment]
Ok, I've implemented the function of @salmon-charles to return the parsed attributes and seems to work fine. This allows to do thinks like:

```js
ns.read('file.sketch').then(sketch => {
    sketch.textStyles.forEach((node) => {
            const node = sketch.pages[0].get('MSAttributedString');
            console.log(node.attributes); //returns the parsed attributes.
    })
})
```

The changes are in the `feature/text-parse` branch, feel free to test it and give feedback.
We need also the opposite function (create the BASE64 encoded Binary Plists) in order to allow edit the values and be able to save it in the sketch file. Currently it is read-only.

**Comment by BenjaminDobler** [comment]
Since sketch 50 the string is not encoded anymore! This should make string reading writing quite easy!

**Comment by oscarotero** [comment]
Good news!
I've released v0.13.0 removing the attribute parsing feature because is not needed anymore.
Now you can get or edit any text value like this: https://github.com/oscarotero/node-sketch/blob/master/demos/style-inventory-generator/index.js#L30
